File: packages/compiler/src/transforms/vOnce.ts
import { findProp } from '../utils'
import type { NodeTransform } from '../transform'

export const transformVOnce: NodeTransform = (node, context) => {
  if (node.type === 'JSXElement' && findProp(node, 'v-once')) {
    context.inVOnce = true
  }
}

File: packages/compiler-rs/src/transform/v_once.rs
use napi::Either;
use oxc_ast::ast::JSXChild;

use crate::{
  ir::index::BlockIRNode,
  transform::{ContextNode, TransformContext},
  utils::utils::find_prop,
};

pub fn transform_v_once<'a>(
  context_node: &'a mut ContextNode<'a>,
  context: &'a TransformContext<'a>,
  _: &'a mut BlockIRNode<'a>,
  _: &'a mut ContextNode<'a>,
) -> Option<Box<dyn FnOnce() + 'a>> {
  let Either::B(node) = context_node else {
    return None;
  };
  if let JSXChild::Element(node) = &node
    && find_prop(node, Either::A(String::from("v-once"))).is_some()
  {
    *context.in_v_once.borrow_mut() = true;
  }
  None
}
